### PR TITLE
Handle gzip responses

### PR DESCRIPTION
**Pull request by kjanoudi**
In connection with https://github.com/blacklane/kiev/issues/13

### CODE REVIEW

File: lib/kiev/rack/request_logger.rb [hunk 1]
```diff
@@ -117,16 +117,29 @@ def form_data(request:, began_at:, status:, env:, body:, response:, exception:)
           data[:body] = full_body.join
         end
 
+        if data[:body] && !data[:body].empty? && response_gzipped?(response)
```
Comment (lipanski) [inline lib/kiev/rack/request_logger.rb]: this code should be wrapped inside the `if config.log_response_body_condition.call(request, response)` condition above.

File: lib/kiev/rack/request_logger.rb [hunk 2]
```diff
@@ -117,16 +117,29 @@ def form_data(request:, began_at:, status:, env:, body:, response:, exception:)
...
           data[:error_class] = exception.class.name
           data[:error_message] = exception.message[0..5000]
           data[:error_backtrace] = Array(exception.backtrace).join(NEW_LINE)[0..5000]
         end
-
         data
       end
 
+      def response_gzipped?(response)
+        response && response.headers && response.headers.is_a?(Hash) && response.headers["Content-Encoding"] == "gzip"
```
Comment (lipanski) [inline lib/kiev/rack/request_logger.rb]: this code is a bit defensive. better said, can `response` be anything else but `Rack::Response` at this point?

if not, then you can reduce this to `response.headers["Content-Encoding"] == "gzip"`

File: lib/kiev/version.rb [hunk 3]
```diff
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
 
 module Kiev
-  VERSION = "4.0.0"
+  VERSION = "4.1.1"
```
Comment (lipanski) [inline lib/kiev/version.rb]: please revert this change, we handle this once the PR is merged into master.

**Comment by lipanski** [comment]
@kjanoudi thanks for your PR. I left a few comments. please add some tests as well - ideally something that tests both the happy path but also the gzip error.

**Comment by kjanoudi** [comment]
@lipanski thanks for the feedback, please review the updated PR

**Comment by lipanski** [comment]
@kjanoudi great, can you please run `rubocop -a` to fix the style errors in your code? (note that there are style errors in other parts of the code which we can fix on our end, once this gets into master)

File: lib/kiev/rack/request_logger.rb [hunk 4]
```diff
@@ -115,6 +115,15 @@ def form_data(request:, began_at:, status:, env:, body:, response:, exception:)
             full_body << str
           end
           data[:body] = full_body.join
+          if data[:body] && !data[:body].empty? && response.headers["Content-Encoding"] == "gzip"
+            begin
+              sio = StringIO.new(data[:body])
+              gz = Zlib::GzipReader.new(sio)
```
Comment (nTraum) [inline lib/kiev/rack/request_logger.rb]: I'm wondering why we don't have to `require 'zlib'` here. Is this implicitly loaded by another gem / require statement already? If yes, shall we add it explicitly here too so we are not betting on the implicit `include'`?

Comment (kjanoudi) [reply lib/kiev/rack/request_logger.rb]: I added the explicit require

**Comment by kjanoudi** [comment]
Ready for review again @nTraum @lipanski

**Comment by lipanski** [comment]
Released as [4.2.0](https://rubygems.org/gems/kiev/versions/4.2.0). Thanks for your contribution! :1st_place_medal:
